data/tutorials/language/0it_06_hof.md

```diff
@@ -0,0 +1,930 @@
+Array.sort Int.compare int_array;;
+
+List.sort String.compare ["z";"b";"a"];;
+
+List.sort Bool.compare [true;false;false];;
+```
+
+### Binding
+
+One last common higher-order pattern in functional programming is the ability to _join_ data from within. For historical reasons, this is normally called a _bind_.
+
+For example, if we have a list, and map over it with a function that returns a list, then we'll have a list of lists. Sometimes we want this, but some times we would rather the new list was _flattened_ instead of _nested_.
+
+To do this with lists we can use the `concat_map` function, which looks like this:
+
+```ocaml=
+let twice x = [x;x];;
+let double ls = List.concat_map twice ls;;
+
+double [1;2;3];; (* [1;1;2;2;3;3] *)
+```
+
+#### Binding as early returns
+
+This same pattern is useful to build chains of functions that _short circuit_ on specific values.
+
+For example, if you had to retrieve a user from a database, and *only if there is a user* try access the user's email, you could use `Option.bind` to short-circuit on the first operation:
+
+```ocaml=
+type user = {
+    email: string option
+}
+
+let get_user () = None ;;
+let get_email user = user.email ;;
+
+let email = Option.bind (get_user ()) get_email ;;
+```
+
+In this example, because our `get_user ()` call returns None, `get_email` will never be called. Only if `get_user` returns `Some user` does `get_email` be called.
+
+This also applies to result values:
+
+```ocaml=
+type user = {
+    email: string
+}
+
+let get_user () = Error `no_database ;;
+let get_email user = Ok user.email ;;
+
+let email = Result.bind (get_user ()) get_email ;;
+```
+
+The main difference is that in this case `Result.bind` is biased towards `Ok value`, so that if a result value is an `Error reason`, it will short-circuit through the binds and just return the first error that appears.
+
+#### Let-ops
+
+Unfortunately, calling `Result.bind` can be a bit awkward. We can't pipe our value through a series of binds like we can do with calls to `map`. For example, this isn't valid:
+
+```ocaml=
+let email = get_user ()
+           |> Result.bind get_email
+           |> Result.bind extract_domain
+           |> Result.bind validate_domain
+           ;;
+```
+
+Thankfully, OCaml lets us redefine a subset of operators called _let-operators_ that can be used to streamline the use of `bind` calls by making them look very close to normal let-bindings:
+
+```ocaml=
+(* first we will declare out let* operator to be equal to Result.bind *)
+let (let*) = Result.bind
+
+let* user = get_user () in
+let* email = get_email user in
+let* domain = extract_domain email in
+validate_domain domain
+```
+
+This has the advantage of making code a lot more readable, without changing the behavior we've come to expect from bind calls.
+
+#### Async code
+
+Async libraries for OCaml that implement Promises/Futures usually also have a `bind` function that allows you to chain computations.
+
+
+<!-- ## 3. Recipes
```

Comment:
At @sabine's request, I've removed this section since most of the things we could think of were already dealt with above. Still I left the content in case you folks think its best to also have it summarized here below.